cloudferrylib/os/actions/vm_snapshot.py

```diff
@@ -90,37 +115,30 @@ class VmRestore(VmSnapshotBasic):
 
     def run(self, rollback_vars=None, *args, **kwargs):
         LOG.debug("restoring vms from snapshot")
-        snapshot_from_namespace = kwargs.get(self.namespace_variable)
-        compute = self.get_compute_resource()
-        position = self.cloud.position
-        vm_id_targets = ('src_id', 'dst_id') \
-            if position == 'src' else ('dst_id', 'src_id')
+        processing_vms_dst = kwargs.get('processing_vms_dst', [])
+        processing_vms_src = kwargs.get('processing_vms_src', {})
+        dst_compute = self.get_dst_compute_resource()
+        src_compute = self.get_src_compute_resource()
+        vm_id_targets = ('src_id', 'dst_id')
         vms_succeeded = {}
         if rollback_vars:
             vms = rollback_vars.get('vms', [])
             vms_succeeded = {pair_vms[vm_id_targets[0]]: pair_vms[
                 vm_id_targets[1]] for pair_vms in vms}
-        for vm in self.get_list_of_vms():
-            if vm.id in vms_succeeded:
-                LOG.debug("Successfully copied instances")
-                if position == 'src':
-                    LOG.debug("SRC ID %s", vm.id)
-                    LOG.debug("DST ID %s", vms_succeeded[vm.id])
-                else:
-                    LOG.debug("SRC ID %s", vms_succeeded[vm.id])
-                    LOG.debug("DST ID %s", vm.id)
-                continue
-            if vm.id not in snapshot_from_namespace:
-                if position == 'dst':
-                    # delete this vm - we don't have its id in snapshot data
-                    LOG.debug("VM %s will be deleted on %s", vm.id, position)
-                    compute.delete_vm_by_id(vm.id)
-            elif vm.status != snapshot_from_namespace.get(vm.id):
-                LOG.debug("Status of %s is changed from %s to %s on %s",
-                          vm.id, vm.status, snapshot_from_namespace.get(vm.id),
-                          position)
-                # reset status of vm
-                compute.change_status(
-                    snapshot_from_namespace.get(vm.id),
-                    instance_id=vm.id)
+        for src_id, dst_id in vms_succeeded.items():
+            LOG.debug("Successfully copied instances")
+            LOG.debug("SRC ID %s", src_id)
+            LOG.debug("DST ID %s", dst_id)
+        for processing_vm in processing_vms_dst:
+            LOG.debug("VM %s will be deleted on %s", processing_vm, "DST")
+            dst_compute.delete_vm_by_id(processing_vm)
```

Comment:
You should only delete instance if it's in non-active state, if VM is active, it should not be deleted